File: python/koans/14_no_starve_mutex/test_no_starve_mutex.py
import inspect
import time

from koan_utils import (
    ThreadRunner,
    EventLog,
    assert_blocks,
    assert_completes,
    jitter,
)

from no_starve_mutex import NoStarveMutex

FORBIDDEN = (
    "threading.Semaphore",
    "threading.BoundedSemaphore",
    "threading.Lock",
    "threading.RLock",
    "threading.Condition",
    "threading.Event",
    "threading.Barrier",
)


def assert_honor_rule():
    src = inspect.getsource(NoStarveMutex)
    for token in FORBIDDEN:
        assert token not in src, (
            f"honor rule: {token} found inside NoStarveMutex — build it from "
            "WeakSemaphore instances and plain ints only"
        )


def test_honor_rule_and_basic_acquire_release():
    assert_honor_rule()
    m = NoStarveMutex()
    assert_completes(m.acquire, timeout=2, msg="an uncontended acquire must not block")
    m.release()
    assert_completes(m.acquire, timeout=2, msg="re-acquire after release must succeed")
    m.release()


def test_mutual_exclusion():
    """A racy read-modify-write must never lose updates under this lock."""
    assert_honor_rule()
    m = NoStarveMutex()
    state = {"value": 0}
    n_threads, n_iters = 4, 50
    runner = ThreadRunner()

    def worker():
        for i in range(n_iters):
            m.acquire()
            temp = state["value"]
            if i % 10 == 0:
                time.sleep(0.0005)  # invite the scheduler to interleave
            state["value"] = temp + 1
            m.release()

    for _ in range(n_threads):
        runner.spawn(worker)
    runner.join_all(timeout=30)
    assert state["value"] == n_threads * n_iters, (
        f"lost updates: expected {n_threads * n_iters}, got {state['value']} — "
        "the critical section is not exclusive"
    )


def test_second_acquire_blocks_until_release():
    m = NoStarveMutex()
    assert_completes(m.acquire, timeout=2, msg="first acquire must not block")
    probe = assert_blocks(
        m.acquire, msg="second acquire must block while the lock is held"
    )
    m.release()
    assert probe.wait(5), "a waiter should get the lock after release"
    m.release()


N_THREADS = 8
LAPS = 150
# Morris bounds overtaking at roughly two "waiting rooms" worth of threads;
# the slack absorbs scheduling noise around the moment intent is recorded.
BOUND = 2 * N_THREADS + 4


def _max_overtakes(events, marked):
    """Worst number of foreign lock grants between the marked thread's
    declared intent (want) and its own grant (got)."""
    want, got = f"want:{marked}", f"got:{marked}"
    worst = count = 0
    in_window = False
    for e in events:
        if e == want:
            in_window, count = True, 0
        elif e == got:
            if in_window:
                worst = max(worst, count)
            in_window = False
        elif in_window and e.startswith("got:"):
            count += 1
    return worst


def test_bounded_overtaking():
    """THE test: with only weak semaphores underneath, nobody may be
    overtaken more than a bounded number of times per acquisition."""
    m = NoStarveMutex()
    log = EventLog()
    runner = ThreadRunner()

    def worker(i):
        for _ in range(LAPS):
            log.record(f"want:{i}")
            m.acquire()
            log.record(f"got:{i}")
            time.sleep(0.0001)  # a sliver of critical-section work
            m.release()
            jitter(0.3)

    for i in range(N_THREADS):
        runner.spawn(worker, i, name=f"P{i}")
    runner.join_all(timeout=45)

    events = log.events()
    assert log.count("got:0") == LAPS
    worst = _max_overtakes(events, marked=0)
    assert worst <= BOUND, (
        f"thread 0 was overtaken {worst} times while waiting for one "
        f"acquisition (bound: {BOUND}) — a weak semaphore is starving it"
    )


def test_progress_under_contention():
    """Heavy contention must not deadlock; every thread finishes its laps."""
    m = NoStarveMutex()
    state = {"count": 0}
    n_threads, laps = 8, 60
    runner = ThreadRunner()

    def worker():
        for _ in range(laps):
            m.acquire()
            state["count"] += 1  # protected by the lock under test
            m.release()
            jitter(0.3)

    for _ in range(n_threads):
        runner.spawn(worker)
    runner.join_all(timeout=30)
    assert state["count"] == n_threads * laps

File: cpp/koans/14_no_starve_mutex/test_no_starve_mutex.cpp
#include "koan_test.hpp"
#include "no_starve_mutex.hpp"

#include <atomic>
#include <memory>
#include <string>
#include <thread>
#include <vector>

using namespace koans;

namespace {

// An increment with a deliberately widened read-modify-write window.
// Relaxed atomics keep the race observable without undefined behavior.
class RacyCounter {
  public:
    void increment(int nap_every, int i) {
        int temp = value_.load(std::memory_order_relaxed);
        if (i % nap_every == 0)
            std::this_thread::sleep_for(500us);  // invite an interleaving
        value_.store(temp + 1, std::memory_order_relaxed);
    }

    int value() const { return value_.load(); }

  private:
    std::atomic<int> value_{0};
};

// A sub-millisecond jitter for high-churn loops.
void tiny_jitter() {
    thread_local std::mt19937 rng{std::random_device{}()};
    std::uniform_int_distribution<int> dist(0, 300);
    std::this_thread::sleep_for(std::chrono::microseconds(dist(rng)));
}

constexpr int kThreads = 8;
constexpr int kLaps = 250;
// Morris bounds overtaking at roughly two "waiting rooms" worth of threads;
// the slack absorbs scheduling noise around the moment intent is recorded.
constexpr int kBound = 2 * kThreads + 4;

// Worst number of foreign lock grants between the marked thread's declared
// intent (want) and its own grant (got).
int max_overtakes(const std::vector<std::string>& events, int marked) {
    const std::string want = "want:" + std::to_string(marked);
    const std::string got = "got:" + std::to_string(marked);
    int worst = 0, count = 0;
    bool in_window = false;
    for (const auto& e : events) {
        if (e == want) {
            in_window = true;
            count = 0;
        } else if (e == got) {
            if (in_window) worst = std::max(worst, count);
            in_window = false;
        } else if (in_window && e.rfind("got:", 0) == 0) {
            ++count;
        }
    }
    return worst;
}

}  // namespace

// A racy read-modify-write must never lose updates under this lock.
KOAN_TEST(mutual_exclusion) {
    NoStarveMutex m;
    RacyCounter counter;
    constexpr int n_threads = 4, n_iters = 50;
    ThreadRunner runner;
    for (int t = 0; t < n_threads; ++t) {
        runner.spawn([&] {
            for (int i = 0; i < n_iters; ++i) {
                m.acquire();
                counter.increment(10, i);
                m.release();
            }
        });
    }
    runner.join_all(30000ms);
    KOAN_ASSERT_MSG(counter.value() == n_threads * n_iters,
                    "lost updates: expected " +
                        std::to_string(n_threads * n_iters) + ", got " +
                        std::to_string(counter.value()) +
                        " — the critical section is not exclusive");
}

KOAN_TEST(second_acquire_blocks_until_release) {
    auto m = std::make_shared<NoStarveMutex>();
    assert_completes([m] { m->acquire(); }, 2000ms, "the first acquire");
    auto probe = assert_blocks([m] { m->acquire(); }, 300ms,
                               "a second acquire (lock is held)");
    m->release();
    probe.assert_completed(5000ms, "the blocked acquire after release");
    m->release();
}

// THE test: with only weak semaphores underneath, nobody may be overtaken
// more than a bounded number of times per acquisition.
KOAN_TEST(bounded_overtaking) {
    NoStarveMutex m;
    EventLog log;
    ThreadRunner runner;
    for (int i = 0; i < kThreads; ++i) {
        runner.spawn(
            [&, i] {
                const std::string want = "want:" + std::to_string(i);
                const std::string got = "got:" + std::to_string(i);
                for (int lap = 0; lap < kLaps; ++lap) {
                    log.record(want);
                    m.acquire();
                    log.record(got);
                    std::this_thread::sleep_for(100us);  // sliver of CS work
                    m.release();
                    tiny_jitter();
                }
            },
            "P" + std::to_string(i));
    }
    runner.join_all(15000ms);

    KOAN_ASSERT_EQ(log.count("got:0"), static_cast<std::size_t>(kLaps));
    int worst = max_overtakes(log.events(), 0);
    KOAN_ASSERT_MSG(worst <= kBound,
                    "thread 0 was overtaken " + std::to_string(worst) +
                        " times while waiting for one acquisition (bound: " +
                        std::to_string(kBound) +
                        ") — a weak semaphore is starving it");
}

// Heavy contention must not deadlock; every thread finishes its laps.
KOAN_TEST(progress_under_contention) {
    NoStarveMutex m;
    int count = 0;  // protected by the lock under test
    constexpr int n_threads = 8, laps = 60;
    ThreadRunner runner;
    for (int t = 0; t < n_threads; ++t) {
        runner.spawn([&] {
            for (int i = 0; i < laps; ++i) {
                m.acquire();
                ++count;
                m.release();
                tiny_jitter();
            }
        });
    }
    runner.join_all(30000ms);
    KOAN_ASSERT_EQ(count, n_threads * laps);
}
